File: Java/Sorting Algorithms/InsertionSort.java
public class InsertionSort extends SortingAlgorithm {

  private static void insert(int[] array, int element, int limit) {
    for (int i=0;i<limit;i++) {
      if (array[i] > element) {
        int tmp = array[i];
        array[i] = element;
        element = tmp;
      }
    }
  }

  public void sort(int[] array) {
    for (int i=0;i<array.length;i++) {
      insert(array, array[i], i);
    }
  }

}

File: Ocaml/InsertionSort.ml
let rec reduce l a f = match l with
  | h::t -> reduce t (f a h) f
  | [] -> a

let rec insert i l = match l with
  | [] -> [i]
  | h::t -> if h > i then i::l else h::(insert i t)

let sort l = reduce l [] insert
